SConstruct

```diff
@@ -475,24 +515,55 @@ if selected_platform in platform_list:
         env["verbose"] = methods.get_cmdline_bool("verbose", True)
         env["warnings"] = ARGUMENTS.get("warnings", "extra")
         env["werror"] = methods.get_cmdline_bool("werror", True)
-        if env["tools"]:
-            env["tests"] = methods.get_cmdline_bool("tests", True)
+        env["tests"] = methods.get_cmdline_bool("tests", True)
     if env["production"]:
         env["use_static_cpp"] = methods.get_cmdline_bool("use_static_cpp", True)
         env["debug_symbols"] = methods.get_cmdline_bool("debug_symbols", False)
         # LTO "auto" means we handle the preferred option in each platform detect.py.
         env["lto"] = ARGUMENTS.get("lto", "auto")
-        if not env["tools"] and env["target"] == "debug":
-            print(
-                "WARNING: Requested `production` build with `tools=no target=debug`, "
-                "this will give you a full debug template (use `target=release_debug` "
-                "for an optimized template with debug features)."
-            )
 
     # Must happen after the flags' definition, as configure is when most flags
     # are actually handled to change compile options, etc.
     detect.configure(env)
 
+    # Set optimize and debug_symbols flags.
+    # Needs to happen after configure to have `env.msvc` defined.
```

Comment:
@Faless I started from https://github.com/godotengine/godot-cpp/blob/master/tools/targets.py but as I started syncing with the platform-specific `detect.py` I had to change some things which should likely be unified further in godot-cpp:

- Removed `0` to `3` flags as `3` isn't valid for MSVC, and IMO the use for these is limited. I think users can specify what they need with `CCFLAGS` if they need to override what we have (we could add `optimize=custom` for this if need be, to make sure we don't end up adding `-O0` from `optimize=none` after user-specified `CCFLAGS`).
- Removed `auto` as the logic is more complex than this, instead we handle it in `SConstruct` when handling the targets.
- Removed forcing debug symbols on `-dev` (ex `debug`) builds. There's a sane default now, but I think we should still give precedence to the `debug_symbols` option if specified.
- MSVC:
  * Removed `/Z7` for builds without debug symbols. Copy paste mistake I believe?
  * Added `/FS` for `debug_symbols=yes` as this was used in `platform/windows/detect.py`. Didn't assess what it is/does.
  * `size` optimization: the proper flag for this is `/O1`: https://learn.microsoft.com/en-us/cpp/build/reference/o1-o2-minimize-size-maximize-speed?view=msvc-170. `/Os` is a subset of `/O1` but the equivalent to GCC `-Os` is `/O1` as far as I know (this also makes a case for not exposing `0` - `3` flags as `-O1` and `/O1` are quite different).
  * Added `env.Append(LINKFLAGS=["/OPT:REF"])` which was in the Windows `detect.py`.
- GCC/Clang:
  * Added `speed-debug` for `-O2` on what used to be `release_debug` builds, as that's what we used there (apparently to keep potential backtraces somewhat usable).

Reply:
I think all the changes are good, and I agree having a `optmize=custom` that requires you to set CCFLAGS seems reasonable.

`/FS` seems to be a partial workaround to windows having issues with multiprocessing.
They say it helps though it doesn't actually fix all the problems: https://learn.microsoft.com/en-us/cpp/build/reference/fs-force-synchronous-pdb-writes?view=msvc-170
Seems safe to use.

Reply:
I added `optimize=custom` in the latest iteration.